File: python-models/src/soma_models/torch/v1/modules/encoder.py
import torch
import torch.nn as nn
from dataclasses import dataclass
from soma_models.config import V1_NUM_LAYERS, V1_EMBEDDING_DIM, V1_PWFF_HIDDEN_DIM, V1_NUM_HEADS, V1_MAX_WAVELENGTH, V1_SCALE_FACTOR
from soma_models.torch.v1.modules.layer import Layer, LayerConfig


@dataclass
class EncoderConfig:
    dropout_rate: float
    embedding_dim: int = V1_EMBEDDING_DIM
    pwff_hidden_dim: int = V1_PWFF_HIDDEN_DIM
    num_layers: int = V1_NUM_LAYERS
    num_heads: int = V1_NUM_HEADS
    max_wavelength: float = V1_MAX_WAVELENGTH
    scale_factor: float = V1_SCALE_FACTOR


class Encoder(nn.Module):
    def __init__(self, config: EncoderConfig):
        super().__init__()
        self.layers = nn.ModuleList([
            Layer(LayerConfig(
                dropout_rate=config.dropout_rate,
                embedding_dim=config.embedding_dim,
                pwff_hidden_dim=config.pwff_hidden_dim,
                num_heads=config.num_heads,
                max_wavelength=config.max_wavelength,
                scale_factor=config.scale_factor,
            ))
            for _ in range(config.num_layers)
        ])

    def forward(self, input: torch.Tensor, positions: torch.Tensor, attn_mask: torch.Tensor):
        x = input
        for layer in self.layers:
            x = layer(x, positions, attn_mask)
        return x

File: models/src/v1/modules/encoder.rs
use burn::{
    config::Config,
    module::Module,
    nn::Initializer,
    tensor::{Int, Tensor, backend::Backend},
};

use crate::v1::{
    V1_EMBEDDING_DIM, V1_MAX_WAVELENGTH, V1_NUM_HEADS, V1_NUM_LAYERS, V1_PWFF_HIDDEN_DIM,
    V1_SCALE_FACTOR, modules::layer::Layer,
};

#[derive(Config, Debug)]
pub struct EncoderConfig {
    /// The size of the input and output features.
    #[config(default = "V1_EMBEDDING_DIM")]
    pub embedding_dim: usize,
    /// The size of the hidden inner pwff features.
    #[config(default = "V1_PWFF_HIDDEN_DIM")]
    pub pwff_hidden_dim: usize,
    /// The number of transformer layers.
    #[config(default = "V1_NUM_LAYERS")]
    pub num_layers: usize,
    /// The number of transformer heads.
    #[config(default = "V1_NUM_HEADS")]
    pub num_heads: usize,
    /// The type of function used to initialize neural network parameters
    #[config(
        default = "Initializer::KaimingUniform{gain:1.0/num_traits::Float::sqrt(3.0), fan_out_only:false}"
    )]
    pub initializer: Initializer,
    /// The max wavelength for RoPE.
    #[config(default = "V1_MAX_WAVELENGTH")]
    pub max_wavelength: f32,
    /// The RoPE scale factor.
    #[config(default = "V1_SCALE_FACTOR")]
    pub scale_factor: f32,
}

#[derive(Module, Debug)]
pub struct Encoder<B: Backend> {
    layers: Vec<Layer<B>>,
}

impl EncoderConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> Encoder<B> {
        let layers =
            (0..self.num_layers).map(|_| Layer::<B>::new(self, device)).collect::<Vec<_>>();
        Encoder { layers }
    }
}
impl<B: Backend> Encoder<B> {
    pub fn forward(
        &self,
        context: Tensor<B, 3>,
        positions: Tensor<B, 2, Int>,
    ) -> Tensor<B, 3> {
        let mut x = context;
        for layer in self.layers.iter() {
            x = layer.forward(x, positions.clone());
        }
        x
    }
}
